### ISSUE TITLE
type[T] generic aren't assignable to TypeForm[T]

### ISSUE THREAD

**Issue by asford** [microsoft/pyright#10359]
`pyright` does allow assignment of a `type[T]` variable to `TypeForm[T]` within a generic function,
which breaks `TypeIs` checkers using the [PEP747 recommended syntax](https://peps.python.org/pep-0747/#typeform-use-cases) when called within generic functions.

See repro case below, which asserts on the expected behavior and has a traceback of the unexpected failure.
Noting, this is run under `enableExperimentalFeatures` but fails with and without this flag.

I believe, from https://github.com/microsoft/pyright/issues/9900, that this is an error,
and the example below would be expected to both type-check and type-narrow correctly.

It's to clear to me, from either [PEP747 combining with type](https://peps.python.org/pep-0747/#combining-with-type)  or other threads, whether the semi-workaround used here...

```python
def is_assignable[T](value: object, typx: TypeForm[T] | type[T]) -> TypeIs[T]
```

... is intended behavior.

### Repro Example

```python
# pyright: enableExperimentalFeatures=true
from typing_extensions import Any, TypeForm, TypeIs, TypeVar, assert_type

T = TypeVar("T")


def isassignable_typeform(value: object, typx: TypeForm[T]) -> TypeIs[T]: ...


def isassignable_type(value: object, typx: type[T]) -> TypeIs[T]: ...


def isassignable_union(value: object, typx: TypeForm[T] | type[T]) -> TypeIs[T]: ...


def test_nongeneric_works(count: int | str):
    # standard constructs using non-generics all work
    # for type, typeform and type|typeform
    #
    # see: https://peps.python.org/pep-0747/#typeform-use-cases
    if isassignable_union(count, int):
        assert_type(count, int)
    else:
        assert_type(count, str)

    if isassignable_typeform(count, int):
        assert_type(count, int)
    else:
        assert_type(count, str)

    if isassignable_type(count, int):
        assert_type(count, int)
    else:
        assert_type(count, str)


def test_generic_typeform_fails(count, typx: type[T]):
    if isassignable_typeform(count, typx):
        # unexpected failure, type[T] is a valid typeform[T]
        # so using typingform[T] should narrow count to T
        assert_type(count, T)


def test_generic_type_works(count, typx: type[T]):
    if isassignable_type(count, typx):
        assert_type(count, T)


def test_generic_typeunion_works(count, typx: type[T]):
    if isassignable_union(count, typx):
        assert_type(count, T)
```

```bash
$ PYRIGHT_PYTHON_FORCE_VERSION=latest pyright --version && pyright demo_typeform_failures.py 
pyright 1.1.399
WARNING: there is a new pyright version available (v1.1.383 -> v1.1.399).
Please install the new version or set PYRIGHT_PYTHON_FORCE_VERSION to `latest`

/home/alexford/ab/main/demo_typeform_failures.py
  /home/alexford/ab/main/demo_typeform_failures.py:38:37 - error: Argument of type "type[T@test_generic_typeform_fails]" cannot be assigned to parameter "typx" of type "TypeForm[T@isassignable_typeform]" in function "isassignable_typeform"
    "type[type]" is not assignable to "type[TypeForm[T@isassignable_typeform]]" (reportArgumentType)
  /home/alexford/ab/main/demo_typeform_failures.py:41:21 - error: "assert_type" mismatch: expected "T@test_generic_typeform_fails" but received "Unknown" (reportAssertTypeFailure)
2 errors, 0 warnings, 0 informations 
```

**Comment by erictraut** [comment]
The narrowed type in this case is `object*` which is effectively an intersection between `object` and `T`. For more details, refer to [this documentation](https://microsoft.github.io/pyright/#/type-concepts-advanced?id=conditional-types-and-type-variables). You can use the value of `count` anywhere that `T` is required. The `assert_type` will fail, however, because it's doing a type equivalence test, and pyright's equivalency logic doesn't consider `object & T` to be the same as `T`.

```python
def test_generic_typeform_fails(count, typx: type[T]):
    if isassignable_typeform(count, typx):
        x: T = count  # This assignment is OK
```

So I think this is working correctly even if it's not quite what you expected.

**Comment by asford** [comment]
Thank you for the clarification, and apologies on the bad repro case.

I'd made a mistake in specifying the `PYRIGHT_PYTHON_FORCE_VERSION=latest` and have edited the report above to include the proper way of specifying the version. This eliminates the unexpected assignment error. 🤦
